File: apps/vinho-ios/Vinho/Core/Services/StatsService.swift
import Foundation
import Supabase

// MARK: - Models
struct WineStats: Decodable {
    let uniqueWines: Int
    let totalTastings: Int
    let uniqueProducers: Int
    let uniqueRegions: Int
    let uniqueCountries: Int
    let favorites: Int
    let averageRating: Double?
    let tastingsLast30Days: Int
    let lastTastingDate: Date?

    enum CodingKeys: String, CodingKey {
        case uniqueWines = "unique_wines"
        case totalTastings = "total_tastings"
        case uniqueProducers = "unique_producers"
        case uniqueRegions = "unique_regions"
        case uniqueCountries = "unique_countries"
        case favorites
        case averageRating = "average_rating"
        case tastingsLast30Days = "tastings_last_30_days"
        case lastTastingDate = "last_tasting_date"
    }
}

struct DisplayStats {
    struct StatItem {
        let value: String
        let label: String
        let subtitle: String
    }

    let wines: StatItem
    let countries: StatItem
    let rating: StatItem
    let recent: StatItem
}

// MARK: - Stats Service
/// Unified stats service for fetching wine statistics
/// Uses the user_wine_stats view for consistency across the app
@MainActor
class StatsService: ObservableObject {
    static let shared = StatsService()

    private let client = SupabaseManager.shared.client
    @Published var currentStats: WineStats?
    @Published var isLoading = false

    private init() {}

    /// Fetch comprehensive wine statistics for the current user
    /// Single source of truth for all stats across the application
    func fetchUserStats() async -> WineStats? {
        guard let _ = try? await client.auth.session.user.id else { return nil }

        isLoading = true
        defer { isLoading = false }

        do {
            let stats: WineStats = try await client
                .from("user_wine_stats")
                .select("*")
                .single()
                .execute()
                .value

            currentStats = stats
            return stats
        } catch {
            print("Failed to fetch user stats: \(error.localizedDescription)")
            return nil
        }
    }

    /// Get formatted stats for display
    func getDisplayStats() async -> DisplayStats? {
        let stats = await fetchUserStats()
        guard let stats = stats else { return nil }

        return DisplayStats(
            wines: DisplayStats.StatItem(
                value: "\(stats.uniqueWines)",
                label: "Wines",
                subtitle: "\(stats.totalTastings) tastings"
            ),
            countries: DisplayStats.StatItem(
                value: "\(stats.uniqueCountries)",
                label: "Countries",
                subtitle: "\(stats.uniqueRegions) regions"
            ),
            rating: DisplayStats.StatItem(
                value: String(format: "%.1f", stats.averageRating ?? 0),
                label: "Avg Rating",
                subtitle: "\(stats.favorites) favorites"
            ),
            recent: DisplayStats.StatItem(
                value: "\(stats.tastingsLast30Days)",
                label: "This Month",
                subtitle: "tastings"
            )
        )
    }
}

// MARK: - Convenience Extensions
extension StatsService {
    /// Quick access to key metrics
    var totalWines: Int { currentStats?.uniqueWines ?? 0 }
    var totalTastings: Int { currentStats?.totalTastings ?? 0 }
    var totalRegions: Int { currentStats?.uniqueRegions ?? 0 }
    var totalCountries: Int { currentStats?.uniqueCountries ?? 0 }
    var favoriteCount: Int { currentStats?.favorites ?? 0 }
    var averageRating: Double { currentStats?.averageRating ?? 0 }
    var recentActivity: Int { currentStats?.tastingsLast30Days ?? 0 }
}

File: apps/vinho-web/lib/stats-service.ts
import { createClient } from "@/lib/supabase";

export interface WineStats {
  uniqueWines: number;
  totalTastings: number;
  uniqueProducers: number;
  uniqueRegions: number;
  uniqueCountries: number;
  favorites: number;
  averageRating: number | null;
  tastingsLast30Days: number;
  lastTastingDate: string | null;
}

/**
 * Unified stats service for fetching wine statistics
 * Uses the user_wine_stats view for consistency across the app
 */
export class StatsService {
  private supabase = createClient();

  /**
   * Fetch comprehensive wine statistics for the current user
   * Single source of truth for all stats across the application
   */
  async fetchUserStats(): Promise<WineStats | null> {
    try {
      const { data, error } = await this.supabase
        .from("user_wine_stats")
        .select("*")
        .single();

      if (error) throw error;

      return {
        uniqueWines: data.unique_wines || 0,
        totalTastings: data.total_tastings || 0,
        uniqueProducers: data.unique_producers || 0,
        uniqueRegions: data.unique_regions || 0,
        uniqueCountries: data.unique_countries || 0,
        favorites: data.favorites || 0,
        averageRating: data.average_rating,
        tastingsLast30Days: data.tastings_last_30_days || 0,
        lastTastingDate: data.last_tasting_date,
      };
    } catch (error) {
      console.error("Failed to fetch user stats:", error);
      return null;
    }
  }

  /**
   * Get formatted stats for display
   */
  async getDisplayStats() {
    const stats = await this.fetchUserStats();
    if (!stats) return null;

    return {
      wines: {
        value: stats.uniqueWines,
        label: "Wines",
        subtitle: `${stats.totalTastings} tastings`,
      },
      countries: {
        value: stats.uniqueCountries,
        label: "Countries",
        subtitle: `${stats.uniqueRegions} regions`,
      },
      rating: {
        value: stats.averageRating?.toFixed(1) || "0.0",
        label: "Avg Rating",
        subtitle: `${stats.favorites} favorites`,
      },
      recent: {
        value: stats.tastingsLast30Days,
        label: "This Month",
        subtitle: "tastings",
      },
    };
  }
}

// Export singleton instance
export const statsService = new StatsService();
